util/retry/retry.go

```diff
@@ -18,107 +18,93 @@
-	Tag         string        // Tag for helpful logging of backoffs
-	Backoff     time.Duration // Default retry backoff interval
-	MaxBackoff  time.Duration // Maximum retry backoff interval
-	Constant    float64       // Default backoff constant
-	MaxAttempts int           // Maximum number of attempts (0 for infinite)
-	UseV1Info   bool          // Use verbose V(1) level for log messages
-	Stopper     *util.Stopper // Optionally end retry loop on stopper signal
+	var delta = r.opts.RandomizationFactor * backoff
+	// Get a random value from the range [backoff - delta, backoff + delta].
+	// The formula used below has a +1 because time.Duration is an int64, and the
+	// conversion floors the float64.
+	return time.Duration(backoff - delta + rand.Float64()*(2*delta+1))
 }
 
-// WithBackoff implements retry with exponential backoff using
-// the supplied options as parameters. When fn returns Continue
-// and the number of retry attempts haven't been exhausted, fn is
-// retried. When fn returns Break, retry ends. As a special case,
-// if fn returns Reset, the backoff and retry count are reset to
-// starting values and the next retry occurs immediately. Returns an
-// error if the maximum number of retries is exceeded or if the fn
-// returns an error.
-func WithBackoff(opts Options, fn func() (Status, error)) error {
-	backoff := opts.Backoff
-	tag := opts.Tag
-	if tag == "" {
-		tag = "invocation"
+// Next returns whether the retry loop should continue, and blocks for the
+// appropriate length of time before yielding back to the caller. If a stopper
+// is present, Next will eagerly return false when the stopper is stopped.
+func (r *Retry) Next() bool {
+	if r.isReset {
+		r.isReset = false
+		return true
 	}
-	for count := 1; true; count++ {
-		status, err := fn()
-		if status == Break {
-			return err
-		}
-		if err != nil && (!opts.UseV1Info || log.V(1) == true) {
-			log.InfoDepth(1, tag, " failed an iteration: ", err)
-		}
-		var wait time.Duration
-		if status == Reset {
-			backoff = opts.Backoff
-			wait = 0
-			count = 0
-			if !opts.UseV1Info || log.V(1) == true {
-				log.InfoDepth(1, tag, " failed; retrying immediately")
-			}
-		} else {
-			if opts.MaxAttempts > 0 && count >= opts.MaxAttempts {
-				return &MaxAttemptsError{opts.MaxAttempts}
-			}
-			if !opts.UseV1Info || log.V(1) == true {
-				log.InfoDepth(1, tag, " failed; retrying in ", backoff)
-			}
-			wait = backoff + time.Duration(rand.Float64()*float64(backoff.Nanoseconds())*retryJitter)
-			// Increase backoff for next iteration.
-			backoff = time.Duration(float64(backoff) * opts.Constant)
-			if backoff > opts.MaxBackoff {
-				backoff = opts.MaxBackoff
-			}
-		}
-		// Wait before retry.
-		select {
-		case <-time.After(wait):
-			// Continue retrying.
-		case <-opts.Stopper.ShouldStop():
-			return util.Errorf("%s retry loop stopped", tag)
-		}
+
+	if r.opts.MaxRetries > 0 && r.currentAttempt == r.opts.MaxRetries {
+		return false
+	}
+
+	// Wait before retry.
```

Comment:
The way this is used in tests means you're going to always wait before actually trying the code in the body of a for loop constructed with a Retry object. This would not be how I'd want it to work. Am I misreading something?